src/write_output_stream.rs

```diff
@@ -0,0 +1,296 @@
+                            if offset < 0 {
+                                return Err(glib::Error::new(
+                                    crate::IOErrorEnum::InvalidArgument,
+                                    "Invalid Argument",
+                                ));
+                            } else {
+                                SeekFrom::Start(offset as u64)
+                            }
+                        }
+                        glib::SeekType::End => SeekFrom::End(offset),
+                        _ => unimplemented!(),
+                    };
+
+                    loop {
+                        match std_error_to_gio_error(write.seek(pos)) {
+                            None => continue,
+                            Some(res) => return res.map(|_| ()),
+                        }
+                    }
+                }
+                _ => Err(glib::Error::new(
+                    crate::IOErrorEnum::NotSupported,
+                    "Truncating not supported",
+                )),
+            }
+        }
+
+        fn can_truncate(&self, _seekable: &crate::Seekable) -> bool {
+            false
+        }
+
+        fn truncate(
+            &self,
+            _seekable: &crate::Seekable,
+            _offset: i64,
+            _cancellable: Option<&crate::Cancellable>,
+        ) -> Result<(), glib::Error> {
+            Err(glib::Error::new(
+                crate::IOErrorEnum::NotSupported,
+                "Truncating not supported",
+            ))
+        }
+    }
+}
+
+glib_wrapper! {
+    pub struct WriteOutputStream(Object<subclass::simple::InstanceStruct<imp::WriteOutputStream>, subclass::simple::ClassStruct<imp::WriteOutputStream>, WriteOutputStreamClass>) @extends crate::OutputStream, @implements crate::Seekable;
+
+    match fn {
+        get_type => || imp::WriteOutputStream::get_type().to_glib(),
+    }
+}
+
+impl WriteOutputStream {
+    pub fn new<R: Write + Send + 'static>(write: R) -> WriteOutputStream {
+        let obj = glib::Object::new(Self::static_type(), &[])
+            .expect("Failed to create write input stream")
+            .downcast()
+            .expect("Created write input stream is of wrong type");
+
+        let imp = imp::WriteOutputStream::from_instance(&obj);
+        *imp.write.borrow_mut() = Some(imp::Writer::Write(Box::new(write)));
+
+        obj
+    }
+
+    pub fn new_seekable<R: Write + Seek + Send + 'static>(write: R) -> WriteOutputStream {
+        let obj = glib::Object::new(Self::static_type(), &[])
+            .expect("Failed to create write input stream")
+            .downcast()
+            .expect("Created write input stream is of wrong type");
+
+        let imp = imp::WriteOutputStream::from_instance(&obj);
+        *imp.write.borrow_mut() = Some(imp::Writer::WriteSeek(Box::new(write)));
+
+        obj
+    }
+
+    // TODO: Need to implement something to get back the inner
+    //pub fn take_inner(&self) -> Box<dyn std::any::Any + 'static> {
```

Comment:
Here and in the `ReadInputStream` we need something to get back the inner part. Should be done somewhat similar to how we do it in cairo already, which would also allow handling of panics in a nicer way.